### PR TITLE
Lease cleanup v1.x

### PR DESCRIPTION
**Pull request by joshua-kim**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java [hunk 1]
```diff
@@ -469,7 +495,10 @@ public KinesisClientLibConfiguration(String applicationName,
                                          boolean validateSequenceNumberBeforeCheckpointing,
                                          String regionName,
                                          long shutdownGraceMillis,
-                                         BillingMode billingMode) {
+                                         BillingMode billingMode,
+                                         long leaseCleanupIntervalMillis,
+                                         long completedLeaseCleanupThresholdMillis,
+                                         long garbageLeaseCleanupThresholdMillis) {
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java]: This is a public constructor. You should keep this intact, deprecate it and introduce a new constructor.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java [hunk 2]
```diff
@@ -836,6 +868,29 @@ public boolean shouldCleanupLeasesUponShardCompletion() {
...
 
+    /**
+     * @return Rate at which to run lease cleanup thread in {@link com.amazonaws.services.kinesis.leases.impl.LeaseCleanupManager}
+     */
+    public long leaseCleanupIntervalMillis() {
+        return leaseCleanupIntervalMillis;
+    }
+
+    /**
+     * @return Threshold for how long a lease can be pending deletion before checking if the lease can be deleted as
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java]: We actually don't have any minimum wait period for lease in queue. We need to specify that completed shards and garbage shards will be checked for cleanup at different intervals.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisShardSyncer.java [hunk 3]
```diff
@@ -788,9 +629,9 @@ static KinesisClientLease newKCLLease(Shard shard) {
      * @return
      */
     static KinesisClientLease newKCLLeaseForChildShard(ChildShard childShard) throws InvalidStateException {
-        final KinesisClientLease newLease = new KinesisClientLease();
+        KinesisClientLease newLease = new KinesisClientLease();
         newLease.setLeaseKey(childShard.getShardId());
-        final List<String> parentShardIds = new ArrayList<>();
+        List<String> parentShardIds = new ArrayList<>();
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisShardSyncer.java]: why finals were removed?

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShardConsumer.java [hunk 4]
```diff
@@ -123,7 +125,8 @@ private static final GetRecordsRetrievalStrategy makeStrategy(KinesisDataFetcher
             IMetricsFactory metricsFactory,
             long backoffTimeMillis,
             boolean skipShardSyncAtWorkerInitializationIfLeasesExist,
-            KinesisClientLibConfiguration config, ShardSyncer shardSyncer, ShardSyncStrategy shardSyncStrategy) {
+            KinesisClientLibConfiguration config, ShardSyncer shardSyncer, ShardSyncStrategy shardSyncStrategy,
+            LeaseCleanupManager leaseCleanupManager) {
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShardConsumer.java]: Let's introduce a single new constructor and deprecate others.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java [hunk 5]
```diff
@@ -153,13 +158,28 @@ public TaskResult call() {
...
+                    // Check if either the shard end ddb persist is successful or
+                    // if childshards is empty. When child shards is empty then either it is due to
+                    // completed shard being reprocessed or we got RNF from service.
+                    // For these cases enqueue the lease for deletion.
+                    if (successfullyCheckpointedShardEnd || CollectionUtils.isNullOrEmpty(childShards)) {
+                        final KinesisClientLease currentLease = leaseCoordinator.getCurrentlyHeldLease(shardInfo.getShardId());
+                        final LeasePendingDeletion leasePendingDeletion = new LeasePendingDeletion(currentLease, shardInfo);
+
+                        if (!leaseCleanupManager.isEnqueuedForDeletion(leasePendingDeletion)) {
+                            leaseCleanupManager.enqueueForDeletion(leasePendingDeletion);
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java:179]: Add a comment here for Jessica to do shard end checkpointing here.

File: src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseCleanupManager.java [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+            cleanedUpGarbageLease = cleanupLeaseForGarbageShard(lease);
+        }
+
+        return new LeaseCleanupResult(cleanedUpCompletedLease, cleanedUpGarbageLease, wereChildShardsPresent,
+                wasResourceNotFound);
+    }
+
+    private Set<String> getChildShardsFromService(ShardInfo shardInfo) {
+        final String iterator = kinesisProxy.getIterator(shardInfo.getShardId(), ShardIteratorType.LATEST.toString());
+        return kinesisProxy.get(iterator, MAX_RECORDS).getChildShards().stream().map(c -> c.getShardId()).collect(Collectors.toSet());
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseCleanupManager.java]: Let's not use MAX_RECORDS as 1. Let's stick with default.

File: src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseCleanupManager.java [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * Helper class to cleanup of any expired/closed shard leases. It will cleanup leases periodically as defined by
+ * {@link KinesisClientLibConfiguration#leaseCleanupIntervalMillis()} upon worker shutdown, following a re-shard event or
+ * a shard expiring from the service.
+ */
+@RequiredArgsConstructor
+@EqualsAndHashCode
+public class LeaseCleanupManager {
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseCleanupManager.java:59]: Ensure this is in sync with latest 2.X changes. Remember there were couple of logging and null check fixes went in.

**Review by ashwing — COMMENTED** [review]
Initial review comments

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java (hunk 1 above)
Comment (joshua-kim) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java]: Will add a new constructor

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisShardSyncer.java (hunk 3 above)
Comment (joshua-kim) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisShardSyncer.java]: ... not sure, will unstage this change...

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java (hunk 2 above)
Comment (joshua-kim) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java]: Will update this java doc

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java (hunk 5 above)
Comment (joshua-kim) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShutdownTask.java:179]: Will add

File: src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseCleanupManager.java (hunk 6 above)
Comment (joshua-kim) [reply src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseCleanupManager.java]: Will remove this limit

File: src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseCleanupManager.java (hunk 7 above)
Comment (joshua-kim) [reply src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseCleanupManager.java:59]: Sure

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java [hunk 8]
```diff
@@ -836,6 +912,29 @@ public boolean shouldCleanupLeasesUponShardCompletion() {
...
 
+    /**
+     * @return Rate at which to run lease cleanup thread in {@link com.amazonaws.services.kinesis.leases.impl.LeaseCleanupManager}
+     */
+    public long leaseCleanupIntervalMillis() {
+        return leaseCleanupIntervalMillis;
+    }
+
+    /**
+     * @return Threshold in millis at which to check if there are any completed leases (leases for shards which have been
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java]: Interval in millis to ...

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java [hunk 9]
```diff
@@ -836,6 +912,29 @@ public boolean shouldCleanupLeasesUponShardCompletion() {
         return cleanupLeasesUponShardCompletion;
     }
 
+    /**
+     * @return Rate at which to run lease cleanup thread in {@link com.amazonaws.services.kinesis.leases.impl.LeaseCleanupManager}
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java]: Interval in millis to run lease cleanup thread...

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java [hunk 10]
```diff
@@ -836,6 +912,29 @@ public boolean shouldCleanupLeasesUponShardCompletion() {
...
+    /**
+     * @return Threshold in millis at which to check if there are any completed leases (leases for shards which have been
+     * closed as a result of a resharding operation) that need to be cleaned up.
+     */
+    public long completedLeaseCleanupThresholdMillis() {
+        return completedLeaseCleanupThresholdMillis;
+    }
+
+    /**
+     * @return Threshold in millis at which to check if there are any garbage leases (leases for shards which no longer
```
Comment (ashwing) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisClientLibConfiguration.java]: Interval in millis to check...

**Review by ashwing — COMMENTED** [review]
LGTM. Minor comments

**Review by ashwing — APPROVED** [review]
lgtm

**Pull request merged**